src/Microsoft.Framework.ConfigurationModel.Json/JsonConfigurationExtension.cs

```diff
@@ -1,13 +1,33 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.Framework.ConfigurationModel.Json;
+using System;
+using System.IO;
+
 namespace Microsoft.Framework.ConfigurationModel
 {
     public static class JsonConfigurationExtension
     {
         public static IConfigurationSourceRoot AddJsonFile(this IConfigurationSourceRoot configuration, string path)
         {
-            configuration.Add(new JsonConfigurationSource(path));
+            return AddJsonFile(configuration, path, false);
+        }
+
+        public static IConfigurationSourceRoot AddJsonFile(this IConfigurationSourceRoot configuration, string path, bool optional)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new ArgumentException(Resources.Error_InvalidFilePath, "path");
+            }
+
+            var fullPath = PathResolver.ResolveAppRelativePath(path);
+
+            if (!optional || File.Exists(fullPath))
```

Comment:
+1. We should throw upfront if a compulsory but non existing configuration is added. And we should also always add the configuration source for an optional configuration whether or not the file exists. This will make sure that for an optional configuration for which the config file is physically added only after the application load we read the configuration on doing a `Configuration.ReLoad()`. So the code should be something like this: 
```
if(!optional && !File.Exists(fullPath))
{
	throw;
}

configuration.Add(new JsonConfigurationSource(path));
```

The `Load` method of the different configuration sources should be aware if the current configuration source is optional and should ignore loading if the file does not exist.